File: src/main/java/org/myfintech/payment/entity/Contract.java
/*
 * Copyright (c) 2024 MyFintech Payment Service
 * All rights reserved.
 *
 * This software is proprietary and confidential. Unauthorized copying of this file,
 * via any medium, is strictly prohibited.
 * @author : Dhanuka Ranasinghe
 * @since : Date: 11/07/2025
 */
package org.myfintech.payment.entity;
import static org.myfintech.payment.entity.EntityColumnConstants.CONTRACT_CONTRACT_NUMBER;
import static org.myfintech.payment.entity.EntityColumnConstants.FOREIGN_KEY_CLIENT_ID;
import static org.myfintech.payment.entity.EntityColumnConstants.TABLE_CONTRACT;

import java.time.OffsetDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true,callSuper = true)
@Entity
@Table(name = TABLE_CONTRACT)
public class Contract extends AbstractEntity<Long> {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = FOREIGN_KEY_CLIENT_ID)
    private Client client;

    @EqualsAndHashCode.Include
    @Column(name = CONTRACT_CONTRACT_NUMBER, nullable = false)
    private String contractNumber;

    public Contract(
            Long id,
            OffsetDateTime createdDate,
            OffsetDateTime modifiedDate,
            Client client,
            String contractNumber) {
        this.setId(id);
        this.setCreatedDate(createdDate);
        this.setModifiedDate(modifiedDate);
        this.client = client;
        this.contractNumber = contractNumber;
    }
}

File: src/main/kotlin/org/myfintech/payment/entity/Contract.kt
/*
 * Copyright (c) 2024 MyFintech Payment Service
 * All rights reserved.
 *
 * This software is proprietary and confidential. Unauthorized copying of this file,
 * via any medium, is strictly prohibited.
 * @author : Dhanuka Ranasinghe
 * @since : Date: 11/07/2025
 */
package org.myfintech.payment.entity

import jakarta.persistence.*
import java.time.OffsetDateTime

@Entity
@Table(name = EntityColumnConstants.TABLE_CONTRACT)
class Contract(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = EntityColumnConstants.FOREIGN_KEY_CLIENT_ID)
    var client: Client,

    @Column(name = EntityColumnConstants.CONTRACT_CONTRACT_NUMBER, nullable = false)
    var contractNumber: String

) : AbstractEntity<Long>() {

    /**
     * Secondary constructor for convenience, often used in testing or mapping.
     */
    constructor(
        id: Long,
        createdDate: OffsetDateTime,
        modifiedDate: OffsetDateTime,
        client: Client,
        contractNumber: String
    ) : this(client, contractNumber) {
        this.id = id
        this.createdDate = createdDate
        this.modifiedDate = modifiedDate
    }

    constructor(
        id: Long,
        client: Client,
        contractNumber: String
    ) : this(client, contractNumber) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Contract) return false
        // Call super for `id` comparison
        if (!super.equals(other)) return false

        // Compare the explicitly included field
        return contractNumber == other.contractNumber
    }

    override fun hashCode(): Int {
        // Start with the superclass's hash code (based on `id`)
        var result = super.hashCode()
        // Combine with the explicitly included field's hash code
        result = 31 * result + contractNumber.hashCode()
        return result
    }
}